### ISSUE TITLE
[java] UnusedImports: False positive if wildcard is used and only static methods

### ISSUE THREAD

**Issue by albfernandez** [pmd/pmd#2016]
**Affects PMD Version:** 
confirmed from 6.9.0 to 6.17.0 and 6.18

**Rule:**
[UnusedImports](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedimports)

**Description:**
False positives from UnusedImports in some specific conditions:
 * Import with wildcard (example: import java.util.*)
 * Using only calls to static methos who take parameters and return in java.lang  (String, Object, etc)

**Code Sample demonstrating the issue:**

```java
package demo; 

import java.util.*;

public class Demo {
    public void testFunction() {
        Objects.toString(null);
    }
}
```

**Running PMD through:** *CLI and Ant*
```
bash run.sh pmd -d src/ -R category/java/bestpractices.xml/UnusedImports -auxclasspath classes
```

```
sep 15, 2019 12:51:53 AM net.sourceforge.pmd.PMD processFiles
ADVERTENCIA: This analysis could be faster, please consider using Incremental Analysis: https://pmd.github.io/pmd-6.17.0/pmd_userdocs_incremental_analysis.html
src/demo/Demo.java:3:	Avoid unused imports such as 'java.util'
```

Curiously, placing the following test code in UnusedImports.xml works fine:

```xml
    <test-code>
    	<description>False positive in Unused imports with Wildcard</description>
    	<expected-problems>0</expected-problems>
    	<code><![CDATA[
package net.sourceforge.pmd.lang.java.rule.bestpractices.unusedimports;
	
import java.util.*;

public class Demo {
	
    public void testFunction() {
        Objects.toString(null);
    }
}    	
    	 ]]></code>    
    </test-code>

```

**Comment by jsotuyod** [comment]
@albfernandez I can't reproduce this, not as a test case nor from CLI with 6.19.0-SNAPSHOT… Any further info you can provide?

**Comment by albfernandez** [comment]
Hi.
I've prepared this script, it download the sample Demo.java from this issue

```bash

mkdir src
mkdir classes
wget -q -O src/Demo.java https://github.com/pmd/pmd/files/3633414/Demo.txt
javac -d classes/ src/Demo.java
bash pmd-bin-6.18.0/bin/run.sh pmd -d src/ -R category/java/bestpractices.xml/UnusedImports -auxclasspath classes
```

I've discovered auxclasspath seems to work only if prefixed with file:// if the class files are in a directory. This command works fine:

```
bash pmd-bin-6.18.0/bin/run.sh pmd -d src/ -R category/java/bestpractices.xml/UnusedImports -auxclasspath file://$(pwd)/classes
```

I've the same issue with Ant, I'll try to fix or reproduce it tomorrow




[Demo.txt](https://github.com/pmd/pmd/files/3633414/Demo.txt)

**Comment by albfernandez** [comment]
Hi

I've done some more researching....
The problem occurs when you set the auxclasspath. If you don't set the auxclasspath it works fine 

This doesn't work (prints a violation)
`bash pmd-bin-6.18.0/bin/run.sh pmd -d src/ -R category/java/bestpractices.xml/UnusedImports -auxclasspath classes`

this works 
`bash pmd-bin-6.18.0/bin/run.sh pmd -d src/ -R category/java/bestpractices.xml/UnusedImports `

The same applies to Ant.

I need to use auxclasspath for other rules to work

**Comment by adangel** [comment]
With type resolution enabled and a correct auxclasspath, we resolve the type of the ASTName node "Object.toString(null)" as "String" - which is the result type of the method call. I think, what we need to do instead is, assign the result type of the method call not to the ASTName, but to the expression:

```
PrimaryExpression  -> Type = java.lang.String
  |- PrimaryPrefix  -> Type = java.util.Objects
     |- Name "Objects.toString" -> Type = java.util.Objects
  |- PrimarySuffix
     |- Arguments
        |- ArgumentList
```

**Comment by albfernandez** [comment]
I've tested it and seems to work fine for me.

Thanks

**Comment by jorn86** [comment]
This is in the 6.20 milestone, does it mean it's fixed in that version? Because I'm using 6.26 and this issue still occurs.

**Comment by adangel** [comment]
@jorn86 Yes, this is what it means. This issue is supposed to be fixed with 6.20.0 already.
If you still encounter this, then you might have a different case that's not covered. Please open a new issue then. Thanks!
